### ISSUE TITLE
[Memory] Significant Peak Memory Spike in Audio Encoder with OmniZip

### ISSUE THREAD

**Issue by Suzukitakeko** [KD-TAO/OmniZip#6]
I am reproducing the results using the WorldSense dataset on NVIDIA RTX A6000s (48GB), which I believe matches the hardware setup described in the paper. While processing some specific video files, I encountered a severe GPU memory spike when the **OmniZip** token compression is enabled.

Such as：
* **Dataset:** WorldSense
* **Video File:** `videos/cYmkYEXB.mp4`
* **Video Duration:** 0h 4m 47.49s

#### Issue Description 

I confirmed that the inference is using the custom modeling implementation provided by this repository (`omnizip.modeling_qwen2_5_omni`), as triggered by the following logic in the inference script:

```python
if method_wrapper == "OmniZip":
    print("#" * 50)
    print("Use OmniZip for Inference")
    print("#" * 50)
    # The issue lies within this custom implementation
    from omnizip.modeling_qwen2_5_omni import Qwen2_5OmniForConditionalGeneration
else:
    from transformers import Qwen2_5OmniForConditionalGeneration

```

When `OmniZip` is active, the `Qwen2_5OmniAudioFlashAttention2` layer triggers `return_logits=True`. My profiling shows that this causes the **Audio Encoder** stage (`get_audio_features`) to consume an excessive amount of peak memory, effectively negating the benefits of Flash Attention.

I profiled the `get_audio_features` function specifically to isolate the issue.

#### Profiling Evidence 

I used the following code to measure the peak memory allocation specifically during the audio encoding phase:

```python
# Profiling Hook
_orig_get_audio_features = self._model.thinker.get_audio_features

def _profiled_get_audio_features(*args, **kwargs):
    torch.cuda.synchronize()
    torch.cuda.reset_peak_memory_stats()
    out = _orig_get_audio_features(*args, **kwargs)
    torch.cuda.synchronize()
    peak = torch.cuda.max_memory_allocated()
    print(
        f"[mem][audio_tower] peakMB={peak/1024/1024:.1f} ",
        flush=True,
    )
    return out

self._model.thinker.get_audio_features = _profiled_get_audio_features

```

**Comparison Results:**

1. **Baseline (Normal Inference / No Logits):**
* Audio Encoder Peak Memory: **~17.6 GB**

<img width="2266" height="258" alt="Image" src="https://github.com/user-attachments/assets/16abf330-c963-4615-963d-c09fabccc196" />

2. **With OmniZip (Requires Logits):**
* Audio Encoder Peak Memory: **~33.2 GB** (+15.6 GB increase!)
* Log output: `Attn Logits Shape: torch.Size([20, 14374, 14374])`

<img width="2202" height="280" alt="Image" src="https://github.com/user-attachments/assets/04a20c3b-b1e6-47ad-b282-e90120e6e5b6" />

#### Root Cause Analysis 

The issue stems from the explicit materialization of the full  attention matrix in `Qwen2_5OmniAudioFlashAttention2` when `return_logits` is True:

```python
if return_logits:
    with torch.no_grad():
        # ...
        # ⚠️ Explicit construction of [Heads, SeqLen, SeqLen] matrix
        attn_logits = torch.matmul(q, k.transpose(-1, -2)) 
        
        # Softmax operation causes peak memory to double (Input + Output tensors)
        attn_logits = attn_logits / (q.shape[-1] ** 0.5)
        attn_logits = F.softmax(attn_logits, dim=-1) 
        # ...

```

For a sequence length of ~14k, the `attn_logits` tensor (FP16) requires ~7.7 GB. The `softmax` operation temporarily requires both the input and output tensors, causing a transient peak of **~15.4 GB**, which matches the observed spike in my logs.

#### Question regarding Paper 

Could you please clarify **the specific methodology used to measure memory consumption in the paper?**

**This implementation detail raises a critical question regarding the memory reduction claims. If the overall peak memory of the pipeline is measured, this massive $O(N^2)$ allocation in the Audio Encoder will completely dominate the memory footprint for examples with long audio tokens (like the one above).Consequently, regardless of how aggressively OmniZip compresses the tokens for the subsequent LLM stage, the overall peak VRAM will remain bottlenecked by the audio encoder's initial memory spike. This makes it mathematically impossible to reproduce the phenomenon reported in the paper, where the peak memory usage steadily decreases as the token compression rate increases.Is there a specific optimization, chunking strategy, or a different configuration used during your official evaluation to avoid this explicit full-matrix calculation?**

**Comment by xxayt** [comment]
In Qwen2.5-Omni, the audio token rate is approximately 25 tokens per second. For a video with a duration of 0h 4m 47.49s, the expected number of audio tokens should be around 7,200, which is roughly half of the reported sequence length (14,374).

If my understanding is correct, the value 14,374 for the audio sequence length may therefore be somewhat unexpected. It would be helpful to clarify whether this number corresponds strictly to audio tokens, or whether it also includes additional tokens introduced by preprocessing steps such as padding, overlapping windows, resampling, or intermediate feature representations.

Could you please confirm how the sequence length 14,374 is obtained in this case?

**Comment by KD-TAO** [comment]
Sorry for the late reply. The data reported in the paper are the inference time and its corresponding memory requirement in the perfilling phase. 

Due to the need to obtain the attention matrix in the Audio Encoder, after testing, we also fing this will cause a surge in memory requirements for some videos.  We are trying to find optimizations to solve this problem. 

Thank you for your attention!

For the second problem, this is due to the fact that Qwen2.5-Omni performs a pool operation after the Audio Encoder to reduce the number of audio tokens, so the number in the encoder is twice the actual calculation.
